### ISSUE TITLE
What Every Computer Scientist Should Know About Floating-Point Arithmetic

### ISSUE THREAD

**Issue by markcmiller86** [betterscientificsoftware/bssw.io#950]
author: @bartlettroscoe 
eb member: @pagrubel 

Seems like a very thorough coverage of the issues...

https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html

**Comment by bernhold** [comment]
I think we should prefer the original: <https://doi.org/10.1145/103162.103163>

Also, we have two BSSw Fellows who have focused on floating point and might be able to help us assemble more resources for a CC on this topic. (Ignacio Laguna and Cindy Rubio-Gonzalez).

Also, it seems sensible to look through David Bailey's bibliography.  He's a well-known expert too, and I expect he would have some articles at a similar level.

**Comment by markcmiller86** [comment]
Good ideas. Am I missing something though...that doi ref doesn't take me to any actual content. Only an abstract and citation info. I didn't see any place to download a pdf.

**Comment by bernhold** [comment]
> that doi ref doesn't take me to any actual content. Only an abstract and citation info. I didn't see any place to download a pdf.

Hmmm, I do.  Just above the abstract, I see a row of icons, with metrics on the left side and various things on the right side, including an online eReader view and a PDF download button.  In this case it appears to be recognizing my ORNL affiliation.  When I tried it from a personal machine, the buttons changed to "Get Access" which allows purchasing the article for those who don't otherwise have access.

As a matter of (BSSw) policy, I think we ought to *always* cite the DOI if there is one available as a the canonical version of a resource.  

If the resource is paywalled, we could consider adding an additional link to an open access version, if we know of one.  I think the case of the Oracle documentation is probably rather unusual.  But the similar and more common case would be where we happen to know that a preprint is available in a personal or institutional archive.  On the other hand, readers can usually google on the title/authors and find such things when they are available.  And anything without a DOI should be considered less reliable/stable as a resource.

**Comment by markcmiller86** [comment]
> As a matter of (BSSw) policy, I think we ought to _always_ cite the DOI if there is one available as a the canonical version of a resource.

That seems reasonable. But, the utility and value of a link that just returns the *actual* article cannot be under-stated and, whenever that seems *appropriate* (might need to pin down what we mean by that), I think we should do it. Maybe this is an *authoritative* vs *non-authoritative* reference issue? Maybe we could have a way of indicating that in our CC links with some kind of emoji icon or something?

> On the other hand, readers can usually google on the title/authors and find such things when they are available. 

Yeah, that means *every* reader who is interested winds up spending that time. Having spent many hours searching for articles related to, for example, the Apollo Guidance Computer posts I did, it can really be a time sink. IMHO, best if links are known and available to provide them even if its non-authoritative.

**Comment by markcmiller86** [comment]
Maybe authoritative (doi) vs non-authoritative references could be handled by `wikize_refs.py`. The article footnotes the DOI but in the citations, we provide one or more alternatives known at the time of publication.

**Comment by bernhold** [comment]
For any given closed source article, there may be accessible copies that may be either legitimate or unauthorized.  Most publishers authorize self-archiving of preprints (i.e. not the published version) at the personal or institutional level.  But it is not uncommon to find people posting the published versions in various places, which is technically *not* legal in most cases.  I think it would be better if BSSw could avoid linking to illegal copies of papers.  Figuring out whether they're actually legal can take work on our part and may not be straightforward.  It is simpler to avoid those complications.  I'm not sure the risk on our part is high, as long as we're not actually *hosting* illegitimate copies of articles.  But it is non-zero.

**Comment by bernhold** [comment]
Another resource in this space: <https://herbie.uwplse.org/> "Herbie: Find and Fix Floating Point Problems"

**Comment by bernhold** [comment]
http://fpanalysistools.org/ has some tools and tutorial from BSSwF Ignacio Laguna (and I think with contributions from BSSwF Cindy Rubio-Gonzalez)

**Comment by bartlettroscoe** [comment]
This paper is just a relevant now as it was back in 1991 and in fact is more relevant.  Note:

*  https://docs.nvidia.com/cuda/floating-point/index.html

which states:

> Since the widespread adoption in 1985 of the IEEE Standard for Binary Floating-Point Arithmetic (IEEE 754-1985 [1]) virtually all mainstream computing systems have implemented the standard, including NVIDIA with the CUDA architecture. IEEE 754 standardizes how arithmetic results should be approximated in floating point.

and it references this paper.

But while these systems may all use IEEE representations of floating point numbers, I have experienced many systems and compilers over the years that violated some of the behavioral aspects of the IEEE floating point standard in different ways (especially when turning on strong optimizations).  I will note this in my CC article and why this is important to numerical algorithm developers.

**Comment by bartlettroscoe** [comment]
PR #1226 contains the CC article ready for review.  I re-read over a good bit of article and provided a compressed summary of the important points and hopefully provided enough enticement to get people to read it.

As I stated in the BSSw.io editorial board meeting last week, this was one of the top 3 papers I remember reading as a graduate student that had a positive impact on my work.  I don't think my QP solver would have been as robust if I had not read and applied what I learned in this paper.
